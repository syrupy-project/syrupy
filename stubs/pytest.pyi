from typing import (
    Any,
    Callable,
    TypeVar,
)

ReturnType = TypeVar("ReturnType")

def fixture(func: Callable[..., ReturnType]) -> Callable[..., ReturnType]: ...

class Function: ...
class Item: ...
