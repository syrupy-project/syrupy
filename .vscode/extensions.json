{
    "recommendations": [
        "littlefoxteam.vscode-python-test-adapter",
        "ms-python.mypy-type-checker"
    ]
}
