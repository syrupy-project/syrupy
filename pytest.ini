[pytest]
addopts = -p pytester -p syrupy
