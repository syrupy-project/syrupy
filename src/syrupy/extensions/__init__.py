from .amber import AmberSnapshotExtension

DEFAULT_EXTENSION = AmberSnapshotExtension
