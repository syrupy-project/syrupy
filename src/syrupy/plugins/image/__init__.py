from .serializer import AbstractImageSnapshotSerializer


class PNGImageSnapshotSerializer(AbstractImageSnapshotSerializer):
    @property
    def extension(self) -> str:
        return "png"


class SVGImageSnapshotSerializer(AbstractImageSnapshotSerializer):
    @property
    def extension(self) -> str:
        return "svg"
