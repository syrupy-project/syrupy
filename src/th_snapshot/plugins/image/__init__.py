from .io import AbstractImageSnapshotIO
from .serializer import ImageSnapshotSerializer


class PNGImageSnapshotIO(AbstractImageSnapshotIO):
    extension = "png"


class SVGImageSnapshotIO(AbstractImageSnapshotIO):
    extension = "svg"
