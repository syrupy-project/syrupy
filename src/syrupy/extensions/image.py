from typing import (
    TYPE_CHECKING,
    Any,
)

from .single_file import SingleFileSnapshotExtension

if TYPE_CHECKING:
    from syrupy.types import SerializableData


class PNGImageSnapshotExtension(SingleFileSnapshotExtension):
    @property
    def _file_extension(self) -> str:
        return "png"


class SVGImageSnapshotExtension(SingleFileSnapshotExtension):
    @property
    def _file_extension(self) -> str:
        return "svg"

    def serialize(self, data: "SerializableData", **kwargs: Any) -> bytes:
        return str(data).encode("utf-8")
