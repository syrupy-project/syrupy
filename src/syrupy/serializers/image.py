from typing import TYPE_CHECKING

from .raw_single import RawSingleSnapshotSerializer


if TYPE_CHECKING:
    from syrupy.types import SerializableData


class PNGImageSnapshotSerializer(RawSingleSnapshotSerializer):
    @property
    def file_extension(self) -> str:
        return "png"

    def serialize(self, data: "SerializableData") -> bytes:
        if isinstance(data, bytes):
            return data
        raise TypeError(
            "Can not serialize image data. Expected "
            f"'bytes', got '{type(data).__name__}'."
        )


class SVGImageSnapshotSerializer(RawSingleSnapshotSerializer):
    @property
    def file_extension(self) -> str:
        return "svg"

    def serialize(self, data: "SerializableData") -> bytes:
        return str(data).encode("utf-8")
