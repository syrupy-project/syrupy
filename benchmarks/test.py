# type: ignore
import base64
from collections import namedtuple

import pytest

from syrupy.extensions.image import (
    PNGImageSnapshotExtension,
    SVGImageSnapshotExtension,
)

example_cycle_list = [1, 2, 3]
example_cycle_list.append(example_cycle_list)
example_cycle_dict = {"a": 1, "b": 2, "c": 3}
example_cycle_dict.update(d=example_cycle_dict)
example_png = base64.b64decode(
    b"iVBORw0KGgoAAAANSUhEUgAAADIAAAAyBAMAAADsEZWCAAAAG1BMVEXMzMy"
    b"Wlpaqqqq3t7exsbGcnJy+vr6jo6PFxcUFpPI/AAAACXBIWXMAAA7EAAAOxA"
    b"GVKw4bAAAAQUlEQVQ4jWNgGAWjgP6ASdncAEaiAhaGiACmFhCJLsMaIiDAE"
    b"QEi0WXYEiMCOCJAJIY9KuYGTC0gknpuHwXDGwAA5fsIZw0iYWYAAAAASUVO"
    b"RK5CYII="
)
example_svg = (
    '<?xml version="1.0" encoding="UTF-8"?>'
    '<svg viewBox="0 0 50 50" xmlns="http://www.w3.org/2000/svg">'
    '<g><rect width="50" height="50" fill="#fff"/>'
    '<g><g fill="#fff" stroke="#707070">'
    '<rect width="50" height="50" stroke="none"/>'
    '<rect x="0" y="0" width="50" height="50" fill="none"/></g>'
    '<text transform="translate(10 27)" fill="#707070" '
    'font-family="ConsolasForPowerline, Consolas for Powerline" font-size="8">'
    '<tspan x="0" y="0">50 x 50</tspan></text></g></g></svg>'
)
ExampleTuple = namedtuple("ExampleTuple", ["a", "b", "c", "d"])


class ExampleClass:
    a = 1
    b = "2"
    c = example_cycle_list
    d = example_cycle_dict
    _protected_variable = None
    __private_variable = None

    def __init__(self, x=None):
        self.x = x
        self._y = 1
        self.__z = 2

    def public_method(self, a, b=1, *, c, d=None):
        pass

    def _protected_method(self):
        pass

    def __private_method(self):
        pass


test_cases = [
    False,
    True,
    3.5,
    7,
    2 / 6,
    22 / 7,
    "",
    r"Raw string",
    r"Escaped \n",
    r"Backslash \u U",
    "ü•ûüêçüçØ",
    "singleline:",
    "- singleline",
    "line 1\nline 2\nline 3",
    "line 2\nline 2\n  line 3",
    "line 1\r\nline 2\r\nline 3",
    "string with 'quotes'",
    b"Byte string",
    ("this", "is", ("a", "tuple")),
    ExampleTuple(a="this", b="is", c="a", d={"named", "tuple"}),
    {"this", "is", "a", "set"},
    {"contains", "frozen", frozenset({"1", "2"})},
    {"contains", "tuple", (1, 2)},
    {"contains", "namedtuple", ExampleTuple(a=1, b=2, c=3, d=4)},
    {"b": True, "c": "Some text.", "d": ["1", 2], "a": {"e": False}},
    {"b": True, "c": "Some ttext.", "d": ["1", 2], "a": {"e": False}},
    {
        1: True,
        "a": "Some ttext.",
        frozenset({"1", "2"}): ["1", 2],
        ExampleTuple(a=1, b=2, c=3, d=4): {"e": False},
    },
    [1, 2, "string", {"key": "value"}],
    example_cycle_list,
    example_cycle_dict,
    ExampleClass(ExampleClass()),
]


class TestClass:
    def test_method(self, snapshot):
        assert example_svg == snapshot(extension_class=SVGImageSnapshotExtension)
        assert example_svg == snapshot
        assert example_png == snapshot(extension_class=PNGImageSnapshotExtension)

    @pytest.mark.parametrize(
        "actual",
        test_cases,
        ids=lambda x: "",
    )
    def test_parametrized_method(self, snapshot, actual):
        assert actual == snapshot
        snapshot.assert_match(actual)
        assert snapshot == actual

    @pytest.mark.parametrize(
        "actual",
        test_cases,
        ids=lambda x: "",
    )
    class TestNestedParametrizedClass:
        def test_nested_method(self, snapshot, actual):
            assert snapshot == actual
