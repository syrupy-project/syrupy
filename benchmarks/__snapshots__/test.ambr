# name: TestClass.test_method.1
  '<?xml version="1.0" encoding="UTF-8"?><svg viewBox="0 0 50 50" xmlns="http://www.w3.org/2000/svg"><g><rect width="50" height="50" fill="#fff"/><g><g fill="#fff" stroke="#707070"><rect width="50" height="50" stroke="none"/><rect x="0" y="0" width="50" height="50" fill="none"/></g><text transform="translate(10 27)" fill="#707070" font-family="ConsolasForPowerline, Consolas for Powerline" font-size="8"><tspan x="0" y="0">50 x 50</tspan></text></g></g></svg>'
---
# name: TestClass.test_parametrized_method[0]
  False
---
# name: TestClass.test_parametrized_method[0].1
  False
---
# name: TestClass.test_parametrized_method[0].2
  False
---
# name: TestClass.test_parametrized_method[10]
  'ü•ûüêçüçØ'
---
# name: TestClass.test_parametrized_method[10].1
  'ü•ûüêçüçØ'
---
# name: TestClass.test_parametrized_method[10].2
  'ü•ûüêçüçØ'
---
# name: TestClass.test_parametrized_method[11]
  'singleline:'
---
# name: TestClass.test_parametrized_method[11].1
  'singleline:'
---
# name: TestClass.test_parametrized_method[11].2
  'singleline:'
---
# name: TestClass.test_parametrized_method[12]
  '- singleline'
---
# name: TestClass.test_parametrized_method[12].1
  '- singleline'
---
# name: TestClass.test_parametrized_method[12].2
  '- singleline'
---
# name: TestClass.test_parametrized_method[13]
  '
  line 1
  line 2
  line 3
  '
---
# name: TestClass.test_parametrized_method[13].1
  '
  line 1
  line 2
  line 3
  '
---
# name: TestClass.test_parametrized_method[13].2
  '
  line 1
  line 2
  line 3
  '
---
# name: TestClass.test_parametrized_method[14]
  '
  line 2
  line 2
    line 3
  '
---
# name: TestClass.test_parametrized_method[14].1
  '
  line 2
  line 2
    line 3
  '
---
# name: TestClass.test_parametrized_method[14].2
  '
  line 2
  line 2
    line 3
  '
---
# name: TestClass.test_parametrized_method[15]
  '
  line 1
  line 2
  line 3
  '
---
# name: TestClass.test_parametrized_method[15].1
  '
  line 1
  line 2
  line 3
  '
---
# name: TestClass.test_parametrized_method[15].2
  '
  line 1
  line 2
  line 3
  '
---
# name: TestClass.test_parametrized_method[16]
  "string with 'quotes'"
---
# name: TestClass.test_parametrized_method[16].1
  "string with 'quotes'"
---
# name: TestClass.test_parametrized_method[16].2
  "string with 'quotes'"
---
# name: TestClass.test_parametrized_method[17]
  b'Byte string'
---
# name: TestClass.test_parametrized_method[17].1
  b'Byte string'
---
# name: TestClass.test_parametrized_method[17].2
  b'Byte string'
---
# name: TestClass.test_parametrized_method[18]
  <class 'tuple'> (
    'this',
    'is',
    <class 'tuple'> (
      'a',
      'tuple',
    ),
  )
---
# name: TestClass.test_parametrized_method[18].1
  <class 'tuple'> (
    'this',
    'is',
    <class 'tuple'> (
      'a',
      'tuple',
    ),
  )
---
# name: TestClass.test_parametrized_method[18].2
  <class 'tuple'> (
    'this',
    'is',
    <class 'tuple'> (
      'a',
      'tuple',
    ),
  )
---
# name: TestClass.test_parametrized_method[19]
  <class 'ExampleTuple'> (
    a='this',
    b='is',
    c='a',
    d=<class 'set'> {
      'named',
      'tuple',
    },
  )
---
# name: TestClass.test_parametrized_method[19].1
  <class 'ExampleTuple'> (
    a='this',
    b='is',
    c='a',
    d=<class 'set'> {
      'named',
      'tuple',
    },
  )
---
# name: TestClass.test_parametrized_method[19].2
  <class 'ExampleTuple'> (
    a='this',
    b='is',
    c='a',
    d=<class 'set'> {
      'named',
      'tuple',
    },
  )
---
# name: TestClass.test_parametrized_method[1]
  True
---
# name: TestClass.test_parametrized_method[1].1
  True
---
# name: TestClass.test_parametrized_method[1].2
  True
---
# name: TestClass.test_parametrized_method[20]
  <class 'set'> {
    'a',
    'is',
    'set',
    'this',
  }
---
# name: TestClass.test_parametrized_method[20].1
  <class 'set'> {
    'a',
    'is',
    'set',
    'this',
  }
---
# name: TestClass.test_parametrized_method[20].2
  <class 'set'> {
    'a',
    'is',
    'set',
    'this',
  }
---
# name: TestClass.test_parametrized_method[21]
  <class 'set'> {
    'contains',
    'frozen',
    <class 'frozenset'> {
      '1',
      '2',
    },
  }
---
# name: TestClass.test_parametrized_method[21].1
  <class 'set'> {
    'contains',
    'frozen',
    <class 'frozenset'> {
      '1',
      '2',
    },
  }
---
# name: TestClass.test_parametrized_method[21].2
  <class 'set'> {
    'contains',
    'frozen',
    <class 'frozenset'> {
      '1',
      '2',
    },
  }
---
# name: TestClass.test_parametrized_method[22]
  <class 'set'> {
    'contains',
    'tuple',
    <class 'tuple'> (
      1,
      2,
    ),
  }
---
# name: TestClass.test_parametrized_method[22].1
  <class 'set'> {
    'contains',
    'tuple',
    <class 'tuple'> (
      1,
      2,
    ),
  }
---
# name: TestClass.test_parametrized_method[22].2
  <class 'set'> {
    'contains',
    'tuple',
    <class 'tuple'> (
      1,
      2,
    ),
  }
---
# name: TestClass.test_parametrized_method[23]
  <class 'set'> {
    'contains',
    'namedtuple',
    <class 'ExampleTuple'> (
      a=1,
      b=2,
      c=3,
      d=4,
    ),
  }
---
# name: TestClass.test_parametrized_method[23].1
  <class 'set'> {
    'contains',
    'namedtuple',
    <class 'ExampleTuple'> (
      a=1,
      b=2,
      c=3,
      d=4,
    ),
  }
---
# name: TestClass.test_parametrized_method[23].2
  <class 'set'> {
    'contains',
    'namedtuple',
    <class 'ExampleTuple'> (
      a=1,
      b=2,
      c=3,
      d=4,
    ),
  }
---
# name: TestClass.test_parametrized_method[24]
  <class 'dict'> {
    'a': <class 'dict'> {
      'e': False,
    },
    'b': True,
    'c': 'Some text.',
    'd': <class 'list'> [
      '1',
      2,
    ],
  }
---
# name: TestClass.test_parametrized_method[24].1
  <class 'dict'> {
    'a': <class 'dict'> {
      'e': False,
    },
    'b': True,
    'c': 'Some text.',
    'd': <class 'list'> [
      '1',
      2,
    ],
  }
---
# name: TestClass.test_parametrized_method[24].2
  <class 'dict'> {
    'a': <class 'dict'> {
      'e': False,
    },
    'b': True,
    'c': 'Some text.',
    'd': <class 'list'> [
      '1',
      2,
    ],
  }
---
# name: TestClass.test_parametrized_method[25]
  <class 'dict'> {
    'a': <class 'dict'> {
      'e': False,
    },
    'b': True,
    'c': 'Some ttext.',
    'd': <class 'list'> [
      '1',
      2,
    ],
  }
---
# name: TestClass.test_parametrized_method[25].1
  <class 'dict'> {
    'a': <class 'dict'> {
      'e': False,
    },
    'b': True,
    'c': 'Some ttext.',
    'd': <class 'list'> [
      '1',
      2,
    ],
  }
---
# name: TestClass.test_parametrized_method[25].2
  <class 'dict'> {
    'a': <class 'dict'> {
      'e': False,
    },
    'b': True,
    'c': 'Some ttext.',
    'd': <class 'list'> [
      '1',
      2,
    ],
  }
---
# name: TestClass.test_parametrized_method[26]
  <class 'dict'> {
    'a': 'Some ttext.',
    1: True,
    <class 'ExampleTuple'> (
      a=1,
      b=2,
      c=3,
      d=4,
    ): <class 'dict'> {
      'e': False,
    },
    <class 'frozenset'> {
      '1',
      '2',
    }: <class 'list'> [
      '1',
      2,
    ],
  }
---
# name: TestClass.test_parametrized_method[26].1
  <class 'dict'> {
    'a': 'Some ttext.',
    1: True,
    <class 'ExampleTuple'> (
      a=1,
      b=2,
      c=3,
      d=4,
    ): <class 'dict'> {
      'e': False,
    },
    <class 'frozenset'> {
      '1',
      '2',
    }: <class 'list'> [
      '1',
      2,
    ],
  }
---
# name: TestClass.test_parametrized_method[26].2
  <class 'dict'> {
    'a': 'Some ttext.',
    1: True,
    <class 'ExampleTuple'> (
      a=1,
      b=2,
      c=3,
      d=4,
    ): <class 'dict'> {
      'e': False,
    },
    <class 'frozenset'> {
      '1',
      '2',
    }: <class 'list'> [
      '1',
      2,
    ],
  }
---
# name: TestClass.test_parametrized_method[27]
  <class 'list'> [
    1,
    2,
    'string',
    <class 'dict'> {
      'key': 'value',
    },
  ]
---
# name: TestClass.test_parametrized_method[27].1
  <class 'list'> [
    1,
    2,
    'string',
    <class 'dict'> {
      'key': 'value',
    },
  ]
---
# name: TestClass.test_parametrized_method[27].2
  <class 'list'> [
    1,
    2,
    'string',
    <class 'dict'> {
      'key': 'value',
    },
  ]
---
# name: TestClass.test_parametrized_method[28]
  <class 'list'> [
    1,
    2,
    3,
    ...,
  ]
---
# name: TestClass.test_parametrized_method[28].1
  <class 'list'> [
    1,
    2,
    3,
    ...,
  ]
---
# name: TestClass.test_parametrized_method[28].2
  <class 'list'> [
    1,
    2,
    3,
    ...,
  ]
---
# name: TestClass.test_parametrized_method[29]
  <class 'dict'> {
    'a': 1,
    'b': 2,
    'c': 3,
    'd': ...,
  }
---
# name: TestClass.test_parametrized_method[29].1
  <class 'dict'> {
    'a': 1,
    'b': 2,
    'c': 3,
    'd': ...,
  }
---
# name: TestClass.test_parametrized_method[29].2
  <class 'dict'> {
    'a': 1,
    'b': 2,
    'c': 3,
    'd': ...,
  }
---
# name: TestClass.test_parametrized_method[2]
  3.5
---
# name: TestClass.test_parametrized_method[2].1
  3.5
---
# name: TestClass.test_parametrized_method[2].2
  3.5
---
# name: TestClass.test_parametrized_method[30]
  <class 'ExampleClass'> {
    a=1,
    b='2',
    c=<class 'list'> [
      1,
      2,
      3,
      ...,
    ],
    d=<class 'dict'> {
      'a': 1,
      'b': 2,
      'c': 3,
      'd': ...,
    },
    x=<class 'ExampleClass'> {
      a=1,
      b='2',
      c=<class 'list'> [
        1,
        2,
        3,
        ...,
      ],
      d=<class 'dict'> {
        'a': 1,
        'b': 2,
        'c': 3,
        'd': ...,
      },
      x=None,
    },
  }
---
# name: TestClass.test_parametrized_method[30].1
  <class 'ExampleClass'> {
    a=1,
    b='2',
    c=<class 'list'> [
      1,
      2,
      3,
      ...,
    ],
    d=<class 'dict'> {
      'a': 1,
      'b': 2,
      'c': 3,
      'd': ...,
    },
    x=<class 'ExampleClass'> {
      a=1,
      b='2',
      c=<class 'list'> [
        1,
        2,
        3,
        ...,
      ],
      d=<class 'dict'> {
        'a': 1,
        'b': 2,
        'c': 3,
        'd': ...,
      },
      x=None,
    },
  }
---
# name: TestClass.test_parametrized_method[30].2
  <class 'ExampleClass'> {
    a=1,
    b='2',
    c=<class 'list'> [
      1,
      2,
      3,
      ...,
    ],
    d=<class 'dict'> {
      'a': 1,
      'b': 2,
      'c': 3,
      'd': ...,
    },
    x=<class 'ExampleClass'> {
      a=1,
      b='2',
      c=<class 'list'> [
        1,
        2,
        3,
        ...,
      ],
      d=<class 'dict'> {
        'a': 1,
        'b': 2,
        'c': 3,
        'd': ...,
      },
      x=None,
    },
  }
---
# name: TestClass.test_parametrized_method[3]
  7
---
# name: TestClass.test_parametrized_method[3].1
  7
---
# name: TestClass.test_parametrized_method[3].2
  7
---
# name: TestClass.test_parametrized_method[4]
  0.3333333333333333
---
# name: TestClass.test_parametrized_method[4].1
  0.3333333333333333
---
# name: TestClass.test_parametrized_method[4].2
  0.3333333333333333
---
# name: TestClass.test_parametrized_method[5]
  3.142857142857143
---
# name: TestClass.test_parametrized_method[5].1
  3.142857142857143
---
# name: TestClass.test_parametrized_method[5].2
  3.142857142857143
---
# name: TestClass.test_parametrized_method[6]
  ''
---
# name: TestClass.test_parametrized_method[6].1
  ''
---
# name: TestClass.test_parametrized_method[6].2
  ''
---
# name: TestClass.test_parametrized_method[7]
  'Raw string'
---
# name: TestClass.test_parametrized_method[7].1
  'Raw string'
---
# name: TestClass.test_parametrized_method[7].2
  'Raw string'
---
# name: TestClass.test_parametrized_method[8]
  'Escaped \\n'
---
# name: TestClass.test_parametrized_method[8].1
  'Escaped \\n'
---
# name: TestClass.test_parametrized_method[8].2
  'Escaped \\n'
---
# name: TestClass.test_parametrized_method[9]
  'Backslash \\u U'
---
# name: TestClass.test_parametrized_method[9].1
  'Backslash \\u U'
---
# name: TestClass.test_parametrized_method[9].2
  'Backslash \\u U'
---
